google-cloud-clients/google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/v1/MessageDispatcher.java

```diff
@@ -365,86 +357,77 @@ public void processReceivedMessages(List<ReceivedMessage> messages, Runnable don
     }
 
     messagesWaiter.incrementPendingMessages(outstandingBatch.messages.size());
-    synchronized (outstandingMessageBatches) {
-      outstandingMessageBatches.add(outstandingBatch);
-    }
+    outstandingMessageBatches.add(outstandingBatch);
     processOutstandingBatches();
   }
 
-  public void processOutstandingBatches() {
-    while (true) {
-      boolean batchDone = false;
-      Runnable batchCallback = null;
-      OutstandingMessage outstandingMessage;
-      synchronized (outstandingMessageBatches) {
-        OutstandingMessageBatch nextBatch = outstandingMessageBatches.peek();
-        if (nextBatch == null) {
-          return;
-        }
-        outstandingMessage = nextBatch.messages.peek();
-        if (outstandingMessage == null) {
-          return;
-        }
+  private void processOutstandingBatches() {
+    for (OutstandingMessageBatch nextBatch = outstandingMessageBatches.poll();
+        nextBatch != null;
+        nextBatch = outstandingMessageBatches.poll()) {
+      ListIterator<OutstandingMessage> messageIterator = nextBatch.messages.listIterator();
+      while (messageIterator.hasNext()) {
+        OutstandingMessage nextMessage = messageIterator.next();
         try {
           // This is a non-blocking flow controller.
-          flowController.reserve(
-              1, outstandingMessage.receivedMessage().getMessage().getSerializedSize());
+          flowController.reserve(1, nextMessage.receivedMessage.getMessage().getSerializedSize());
         } catch (FlowController.MaxOutstandingElementCountReachedException
             | FlowController.MaxOutstandingRequestBytesReachedException flowControlException) {
-          return;
+          // Nack all messages if flow controlled.  They cannot be handled now.
+          pendingNacks.add(nextMessage.receivedMessage.getAckId());
+          while (messageIterator.hasNext()) {
+            pendingNacks.add(messageIterator.next().receivedMessage.getAckId());
+          }
+          continue;
         } catch (FlowControlException unexpectedException) {
           throw new IllegalStateException("Flow control unexpected exception", unexpectedException);
         }
-        nextBatch.messages.poll(); // We got a hold to the message already.
-        batchDone = nextBatch.messages.isEmpty();
-        if (batchDone) {
-          outstandingMessageBatches.poll();
-          batchCallback = nextBatch.doneCallback;
-        }
+        processOutstandingMessage(nextMessage);
       }
+      nextBatch.doneCallback.run();
```

Comment:
This is really not good idea. You shouldn't call this if msg is flow controlled. Since you are nacking all msg thats flow controlled it will go in to loop of receiving msg , nacking and requesting more msg, utilizing unnecessary bandwidth and will not allow grpc to auto tune the flow control as well. This should only be called when application has actually completed processing all pending received msgs. previous code of calling this on batchDone with correct.